Fix comment trigger against PR branch

andreydevyatkin opened the pull request:
fixes #27715

- Fixes the issue with comment trigger against PR branch
- Fixes the issue with workflow_dispatch event

<img width="1680" alt="Screenshot 2023-07-31 at 20 20 26" src="https://github.com/apache/beam/assets/38127224/2d6b54c5-d27f-46d9-9fde-45ec3d8a9a6d">

<img width="1679" alt="Screenshot 2023-07-31 at 20 19 38" src="https://github.com/apache/beam/assets/38127224/eb361839-1257-4103-a620-9fe39a2c0250">

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

damccorm commented on .github/workflows/beam_PreCommit_Go.yml:
```diff
@@ -37,17 +37,25 @@ concurrency:
-    steps:       
-      - uses: actions/checkout@v3
+    steps:
+      - name: Check out repository code
```
Could we extract this into a shared composite action (like rerun action)?

Abacn replied:
agree, this has become more complex

andreydevyatkin replied:
@damccorm @Abacn It seems there is no concise way to encapsulate the checkout logic in a composite action. Based on the doc, we must checkout the repository before using the custom action https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-an-action-in-the-same-repository-as-the-workflow. The only thing we can do is checkout the custom action code and use this action as the second step:
```
    steps:
      - name: Check out action code
        uses: actions/checkout@v3
        with:
          sparse-checkout: ./.github/actions/checkout
      - name: Check out repository code
        uses: ./.github/actions/checkout
        with:
          comment_phrase: 'Run Go PreCommit'
```
Action:
```
runs:
  using: composite
  steps:
    - name: Check out
      uses: actions/checkout@v3
      if: contains(fromJSON('["push", "pull_request_target", "schedule", "workflow_dispatch"]'), ${{ github.event_name }})
      with:
        ref: ${{ github.event.pull_request.head.sha }}
    - name: Check out if comment phrase
      uses: actions/checkout@v3
      if: ${{ github.event.comment.body }} == ${{ inputs.comment_phrase }}
      with:
        ref: refs/pull/${{ github.event.issue.number }}/head
```
I think this approach adds more unclarity on how to properly use the action compared to the current implementation.

damccorm replied:
Hm, what if we did something like:

```
steps:
- uses: actions/checkout@v3
- uses: ./github/actions/setup-action
```

Where setup-action is an action that (1) checks out the correct ref (we can do this from bash) and (2) calls the rerun action.

Its technically a _little_ less efficient since we're doing 2 checkouts, but we're still only doing one clone and the repo isn't huge, so it should be negligible, and it simplifies the dev workflow.

andreydevyatkin replied:
Encapsulated the checkout logic and the re-run step into the setup action

damccorm reviewed (commented):
Looks like this needs to have merge conflicts fixed

damccorm commented on .github/actions/setup-action/action.yml:
```diff
@@ -0,0 +1,69 @@
+    description: "The GitHub job"
+    required: false
+    default: ''
+  github_current_run_id:
```
Question: I know this was partially pulled from the rerun action bit, but can't we just get a lot of these from the github context object? Do we actually need to have them all passed in as params?

andreydevyatkin replied:
I removed some of the inputs and left only `github_token` (composite actions don't work with secrets directly) and `github_job` (it is widely used by jobs that have matrix values and the it's generated at runtime). As for the remaining, I think it make sense to leave them as inputs for the rerun action in case we ever need to interact with several repositories. We can revisit the rerun action logic some time later.

damccorm commented on .github/actions/setup-action/action.yml:
```diff
@@ -0,0 +1,69 @@
+  using: composite
+  steps:
+    - name: Check out repository code
+      uses: actions/checkout@v3
```
Does having 2 checkout actions actually work? Does it check out 2 copies of the repo or does it actually overwrite the existing one?

I was anticipating this being a bash step like

`git fetch origin ${{ github.event.pull_request.head.sha }} && git checkout ${{ github.event.pull_request.head.sha }}`

damccorm replied:
If it works, that's great, I'm just surprised

andreydevyatkin replied:
Calling `actions/checkout` multiple times it updates/overwrites the same working directory with the latest changes from the specified branch or commit hash each time. It doesn't create multiple copies of the repository (until you explicitly specify it), and instead, it keeps the working directory up-to-date with the latest changes.

damccorm replied:
Ok awesome!

damccorm reviewed (commented):
Just had a couple comments/questions, otherwise this looks good to me. I like how the composite action looks in the main workflows.

damccorm commented:
Could we update https://github.com/apache/beam/blob/master/.github/workflows/README.md#comment-triggering-support and https://github.com/apache/beam/blob/master/.github/workflows/README.md#checkout-step to recommend this instead of the rerun action?

damccorm reviewed (approved):
This LGTM, would you prefer to update the Readme in this PR or defer to the larger update?

andreydevyatkin commented:
> This LGTM, would you prefer to update the Readme in this PR or defer to the larger update?

Updated the README file and updated the setup action in accordance with your comments. Thanks!

Abacn commented:
Looks like this is working, but caused #27854 on PRs that has base branch before this PR merge
